fix(platform): table a11y & keyboard navigation

platon-rov opened the pull request:
#### Please provide a link to the associated issue.

Closes #3841
Closes #4962
Closes #4668

#### Please provide a brief summary of this pull request.

Overall: 
* A11Y issues resolved;
* Keyboard navigation added. 

Issue #3841:
* Fixed.

Issue #4962:
* Issue 1, Keyboard navigation: fixed;
* Issue 2, Column headers: fixed; 
* Issue 3, Description input: fixed;
* Issue 4, Table name, Loading: PR #6061;
* Issue 5, Excess tabs: fixed;
* Issue 6, Multiple selection column label: fixed;
* Issue 7, Keyboard navigation: fixed;
* Issue 8, Focused styles for the selection column: fixed;
* Issue 9, Search field label: fixed;
* Issue 10, Playground: PR #6158;
* Issue 11, Prsentation role: fixed;
* Issue 12, Mobile (pop-in) mode: planned to implement, not fixed;
* Issue 13, Sort dialog, checkboxes focus styles: working;
* Issue 14, Sort dialog (single), checkboxes keyboard navigation: fixed;

Issue #4668:
* Issue 1, Search field label: fixed;
* Issue 2, Search field Submit button: PR #5673;
* Issue 3, Keyboard navigation: fixed;
* Issue 4, Description column label: fixed;
* Issue 5, Sort dialog information: PR #5682;
* Issue 6, Sort dialog, selects keyboard navigation: working;
* Issue 7, Sort dialog, selects value: PR #5385.
* Issue 8, Sort dialog, add/delete buttons focus: working;
* Issue 9, Playground: PR #6158.

Mobile (pop-in) mode - not present, planned to implement.
Custom width, column resize - PR #5879.

#### Please check whether the PR fulfills the following requirements

- [X] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/main/CONTRIBUTING.md
- [X] tests for the changes that have been done
- [X] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist
- [X] Run npm run build-pack-library and test in external application

Documentation checklist:
- [n/a] update `README.md`
- [n/a] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [X] Documentation Examples
- [X] Stackblitz works for all examples

kavya-b commented on libs/platform/src/lib/components/table/components/table-column/table-column.component.ts:
```diff
@@ -132,6 +143,18 @@ export class TableColumnComponent extends TableColumn implements OnInit, OnChang
         this.headerCellTemplate = fdpHeaderCellDef?.templateRef;
     }
 
+    @ContentChild(FdpTableHeader)
```
please add some typedoc for these methods

kavya-b commented on libs/platform/src/lib/components/table/table.component.ts:
```diff
@@ -23,10 +24,11 @@ import {
 import { debounceTime, distinctUntilChanged, filter, map, startWith, switchMap } from 'rxjs/operators';
 
-import { FdDropEvent, RtlService } from '@fundamental-ngx/core/utils';
+import { FdDropEvent, KeyUtil, RtlService } from '@fundamental-ngx/core';
```
shouldn't this be imported from`@fundamental-ngx/core/utils`?

kavya-b commented on libs/platform/src/lib/components/table/table.service.ts:
```diff
@@ -1,16 +1,22 @@
 import { BehaviorSubject, Observable, Subject } from 'rxjs';
 import { skip } from 'rxjs/operators';
 
+import { KeyUtil } from '@fundamental-ngx/core';
```
'@fundamental-ngx/core/utils' import can be used.

kavya-b commented on apps/docs/src/app/platform/component-docs/platform-table/platform-table-docs.component.html:
```diff
@@ -18,7 +18,20 @@
+            Use input property <code>ariaLabel</code> of the <code>fdp-table-cell</code> for specifying how custom column content should be read by screen readers.
+        </li>
+        <li>
+            Use input property <code>[focusable]="true"</code> of the <code>fdp-table-cell</code> if there is some focusable content in the table cell (input, button, link, etc) so the screen reader will read about that user can go into the table cell.
```
```suggestion
            Use input property <code>[focusable]="true"</code> of the <code>fdp-table-cell</code> if there is some focusable content in the table cell (input, button, link, etc) so the screen reader will read about that and user can go into the table cell.
```

kavya-b reviewed (commented):
Looks good, minor comments. 
One question, when I am in RTL mode(RTL switched on) and then navigate through tab+arrow keys, is it expected that pressing right arrow navigates to the left column and pressing left arrow navigates to the right column and vice versa? Is this the expected behavior?

platon-rov commented:
> Looks good, minor comments.
> One question, when I am in RTL mode(RTL switched on) and then navigate through tab+arrow keys, is it expected that pressing right arrow navigates to the left column and pressing left arrow navigates to the right column and vice versa? Is this the expected behavior?

No it's a bug, fixed now.

kavya-b commented on libs/platform/src/lib/components/table/table.service.ts:
```diff
@@ -39,6 +39,8 @@ export class TableService {
         return this._tableCells.get(this._focusTableCellId)?.nativeElement.innerText;
     }
 
+    constructor(private readonly _rtService: RtlService) {}
```
should be `@hidden`

olshansky reviewed (approved):
Great job!

One small comment: since the PR name will be a commit message in the `main` branch the subject shouldn't start from capital letter (`Table a11y & keyboard navigation`)

kavya-b commented:
> Great job!
> 
> One small comment: since the PR name will be a commit message in the `main` branch the subject shouldn't start from capital letter (`Table a11y & keyboard navigation`)

good catch! adding on to @olshansky 's comment, the commit message and PR title should start with a present tense, so this can probably be changed to something like: `fix(platform): fix Table a11y and keyboard navigation`?

kavya-b reviewed (approved):
LGTM!

puru-hk commented:
Issue 4962:
Issue 1: Fixed.
Issue 2: Fixed
**Issue 3- Not Fixed
Now the Description input fields are not accessible using tab key.**
**Issue 4: Partially fixed
The additional information 'loading' is still read on navigating to the table.
The table name is still not read for tables within 'multiple row selection, Making row non-navigatable' headers.**
Issue 5: Fixed
**Issue 6: Issue still persists the same for all the tables.**
Issue 7:Fixed
Issue 8: Fixed
Issue 9: Fixed
Issue 10: Fixed
issue 11: Fixed
Issue 12: Issue still persists horizontal scroll bar are available.
issue 13: Fixed
Issue 14: Fixed

puru-hk commented:
Issue 3841:
The issue is not fixed.
OB: On accessing the table first row, the row and column details is read additional information such as row 1 and colum1 is read and on accessing the table cell data information is not read.
![image](https://user-images.githubusercontent.com/78714518/124061843-403cde80-da4d-11eb-9aa1-fe43e51fff4e.png)
EB: ideally, the table cell details should be read and only relevant column and row details should be read and the data cell information.

platon-rov commented:
> Issue 4962:
> Issue 3 - Not Fixed Now the Description input fields are not accessible using tab key.
> Issue 4: Partially fixed The additional information 'loading' is still read on navigating to the table. The table name is still not read for tables within 'multiple row selection, Making row non-navigatable' headers.
> Issue 6: Issue still persists the same for all the tables.
> Issue 12: Issue still persists horizontal scroll bar are available.

3: From now the behavior is very similar to the SAPUI5 table - when tabbing over the page and focus comes to the table - the first column of the first row is focused. Then, when pressing Tab - going out from the table. When the focus somewhere below the table and pressing Shift + Tab - the first column of the first row gets the focus. For navigation over the table (cells) arrow buttons are intended to use. Pressing F2/Enter navigates to the table cell content if there is focusable content, then pressing Tab navigates to the next closest focusable content in the table.

4: Fixed

6: Fixed

12: We don't have mobile (called pop-in) mode in this component yetю

puru-hk commented:
**Issue 4962**
Issue 3: 
 
OB: On accessing the description column 1st row additional information column1 and row 1 is read and loading information for the rest of the rows are read. 

Expected Behaviour:  column1 and row 1 and loading information should be removed . 

Issue 4: 
OB: no keyboard support provided. 


Issue: 6 

 Multi selectable information is not read and the table name and information about the table is not read.

puru-hk commented:
Issue 12: Tested in real device it is working fine

platon-rov commented:
@Sanjay-1312 Hi, I've processed feedback and updated the description with links, you can review now. If some item looks not fixed check if it's not related to the different issue mentioned in the description, thank you.

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-table/platform-table-examples/platform-table-custom-column-example.component.html:
```diff
@@ -21,7 +21,7 @@
         key="description">
         <fdp-table-header *fdpHeaderCellDef>Description</fdp-table-header>
-        <fdp-table-cell *fdpCellDef="let item">
+        <fdp-table-cell *fdpCellDef="let item" ariaLabelKey='description' focusable="true">
```
```suggestion
        <fdp-table-cell *fdpCellDef="let item" ariaLabelKey="description" focusable="true">
```

Betrozov commented on libs/platform/src/lib/components/table/components/table-p13-dialog/sorting/sorting.component.html:
```diff
@@ -29,6 +28,7 @@
                 [value]="rule.columnKey"
                 (valueChange)="_onRuleColumnKeyChange(rule, $event)"
                 [compact]="true"
+                [ariaLabel]="'Rule #'+ (i + 1) + ', sort by'"
```
should we add i18n here?

platon-rov replied:
I think yes to indicate what exactly this select for.

Betrozov replied:
then please add i18n for ariaLabel

Betrozov commented on libs/platform/src/lib/components/table/table.component.ts:
```diff
@@ -1165,6 +1165,12 @@ export class TableComponent<T = any> extends Table implements AfterViewInit, OnD
         this._calculateCheckedAll();
+
+        if (this._shouldFocusAfterPageHasChanged) {
+            this._cd.detectChanges();
```
could we use markForCheck here instead of detectChanges?

platon-rov replied:
It won't work.

puru-hk commented:
Issue 3841:
OB: On accessing the column row of second and remaining rows the count is not read.
EB: Ideally, the row and column count should be read to the user.
![image](https://user-images.githubusercontent.com/78714518/125911262-ca825f8d-c695-44c9-a9c4-1b5a4f9ddfd1.png)

Betrozov commented on libs/platform/src/lib/components/table/components/table-p13-dialog/sorting/sorting.component.html:
```diff
@@ -45,6 +45,7 @@
                 [value]="rule.direction"
                 (valueChange)="_onRuleDirectionChange(rule, $event)"
                 [compact]="true"
+                [ariaLabel]="'Rule #'+ (i + 1) + ', sort direction'"
```
same

puru-hk commented:
**Issues: 3841**

Hello Colleagues,

I can still able to reproduce the issue.
I have added screenshot for your reference,
Screenshot:
![image](https://user-images.githubusercontent.com/78714518/126111572-3a1ce6c7-9bd4-44d0-875a-da8bb06e6479.png)

puru-hk commented:
**Issue: 4962**

Hello Colleagues,

OB: On accessing the table, the loading information is read and the on accessing the row checkboxes read as has controls.
and on pressing 'F2' button the checkbox can selected.
EB: The loading information shouldn't be read.
![image](https://user-images.githubusercontent.com/78714518/126113772-e5d7774a-69af-4eed-a191-2cc2d87326b9.png)

puru-hk commented:
**Issue: 4668**

Issue 1: Fixed

Issue 2: Fixed

Issue 3: Fixed

Issue 4:
OB: The table two and count is not read on accessing the 2nd row from the field.
EB: The row and column count should be read when every table data cell is accessed.
![image](https://user-images.githubusercontent.com/78714518/126126042-88f5cd3b-2f00-4cdd-9c3d-30710624ecde.png)

Issue 5:
![image](https://user-images.githubusercontent.com/78714518/126126122-c5ac6e3b-e980-4d5a-b4d0-23f4d337d724.png)

OB: On activating the 'Sort' button, the dialog header 'Sort' is not read and the focus is available on the 'Cancel' button.
EB: The dialog name 'Sort' and the focus should be available on the first Active UI element of the dialog.
![image](https://user-images.githubusercontent.com/78714518/126126334-a0385ae4-7baf-4878-9f97-d369df69ec18.png)

issue 6: Fixed

Issue 7:
![image](https://user-images.githubusercontent.com/78714518/126126437-affccc3e-5a1b-4095-805f-4040595bc1a5.png)
![image](https://user-images.githubusercontent.com/78714518/126126469-5985a272-399c-480f-9960-7459ebebc319.png)
OB: The selected name is not read and visible label should be provided

issue 8: Fixed

Issue 9: Fixed


URL:https://deploy-preview-5692--fundamental-ngx.netlify.app/#/platform/table
JAWS 2021

platon-rov commented:
Closing PR until blocker a11y issues not fixed - Dialog (core) & Busy Indicator (core).

platon-rov commented:
Update on blockers.

Dialog: seems to be fixed, appropriate ticket is closed.
Select: https://github.com/SAP/fundamental-ngx/pull/5385
Busy Indicator: https://github.com/SAP/fundamental-ngx/pull/6061 (edited)

platon-rov reviewed (commented):
1 Focus doesn't move out (to the next focusable element under the table) from the first table cell when pressing `tab`.
2 Wrong focused cell styles
![image](https://user-images.githubusercontent.com/20265336/141338205-006084f7-6884-4a42-84d0-cd4ebb476569.png)
3 When having focusable rows (activable, navigatable) or cells (any focusable el in the cell) after pressing tab when the focus is on the last row first table cell gets focus, so we're getting in the loop. 
https://user-images.githubusercontent.com/20265336/141338511-35bfe912-5be9-4629-b9d0-90314283bfc7.mov

mikerodonnell89 reviewed (changes requested):
EDIT: my issue previously mentioned above

dmitry-stepanenko commented on libs/core/src/lib/utils/functions/key-util.ts:
```diff
@@ -36,6 +38,7 @@ const keyMap: Map<number, string[]> = new Map([
     [ESCAPE, ['Escape', 'Esc']],
     [DELETE, ['Delete', 'Del']],
     [ENTER, ['Enter']],
+    [MAC_ENTER, ['Return']],
```
wouldn't it be reasonable to combine these 2? I'm not sure it will be ever a case when we need to track these 2 keys apart

aberikashvili replied:
Removed MAC_ENTER

dmitry-stepanenko commented on e2e/wdio/core/tests/vertical-navigation.e2e-spec.ts:
```diff
@@ -11,7 +11,7 @@ import {
 } from '../../driver/wdio';
 
-describe('Vertical navigation component tests', function () {
+describe('Vertical navigation component tests', function (): void {
```
I guess, for consistency it would be better to just make it an arrow function instead. It won't complain about return type
```suggestion
describe('Vertical navigation component tests', () => {
```

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/components/table-p13-dialog/sorting/sorting.component.ts:
```diff
@@ -132,6 +132,11 @@ export class P13SortingDialogComponent implements Resettable {
         return rule.columnKey;
     }
 
+    _getAriaLabel(rule: ValidatedSortRule): string {
```
typedoc

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/components/table-p13-dialog/sorting/sorting.component.ts:
```diff
@@ -132,6 +132,11 @@ export class P13SortingDialogComponent implements Resettable {
     }
 
+    _getAriaLabel(rule: ValidatedSortRule): string {
+        const columnLabel = this.columns.find((col) => col.key === rule.columnKey)?.label;
```
Can you calculate these values to a map on component init? Since this function is called from UI, it will run `find` on each CD cycle

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/directives/table-cell-selectable.directive.ts:
```diff
@@ -1,19 +1,106 @@
+    _onKeyDown(event: KeyboardEvent): void {
+        if (
+            KeyUtil.isKeyCode(event, F2) ||
+            (KeyUtil.isKeyCode(event, [ENTER, MAC_ENTER]) && !this._tableService.isFocusInsideTableCell)
```
can you swap `isKeyCode` and `!this._tableService.isFocusInsideTableCell` conditions please? Not a big performance issue, but still nice to have

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/directives/table-cell-selectable.directive.ts:
```diff
@@ -1,19 +1,106 @@
+            KeyUtil.isKeyCode(event, F2) ||
+            (KeyUtil.isKeyCode(event, [ENTER, MAC_ENTER]) && !this._tableService.isFocusInsideTableCell)
+        ) {
+            this._tableService.processFocusInsideCell(this.navigationId);
```
at this point you can exit the function, it won't catch any other keycodes below

aberikashvili replied:
hm.. how? I don't understand your point at all

dmitry-stepanenko replied:
```suggestion
            this._tableService.processFocusInsideCell(this.navigationId);
            return;
```

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/directives/table-cell-selectable.directive.ts:
```diff
@@ -1,19 +1,106 @@
+        }
+
+        const arrowKeyCodes = [UP_ARROW, DOWN_ARROW, LEFT_ARROW, RIGHT_ARROW];
+        if (KeyUtil.isKeyCode(event, arrowKeyCodes) && document.activeElement === this._elRef.nativeElement) {
```
same as above, please swap the conditions

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/table.component.ts:
```diff
@@ -1979,6 +1992,52 @@ export class TableComponent<T = any> extends Table implements AfterViewInit, OnD
+        this._subscriptions.add(
+            fromEvent(this.tableContainer.nativeElement, 'keydown').subscribe((event: KeyboardEvent) => {
+                if (this.loading) {
+                    event.preventDefault();
```
this can potentially cause issues with table controls, that are outside the table body. Is it required to call `preventDefault`?

aberikashvili replied:
Removed

dmitry-stepanenko commented on libs/platform/src/lib/table/table.service.ts:
```diff
@@ -226,9 +263,126 @@ export class TableService {
+        const [rowIndex, colIndex] = currCellNavId.split(',');
+        let nextCellId: string;
+
+        if (KeyUtil.isKeyCode(event, UP_ARROW)) {
+            nextCellId = `${+rowIndex - 1},${colIndex}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, DOWN_ARROW)) {
+            nextCellId = `${+rowIndex + 1},${colIndex}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, LEFT_ARROW)) {
+            nextCellId = `${rowIndex},${+colIndex - (this._rtl ? -1 : 1)}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, RIGHT_ARROW)) {
+            nextCellId = `${rowIndex},${+colIndex + (this._rtl ? -1 : 1)}`;
+        }
```
these conditions are exclusive. Use `else if`

aberikashvili replied:
Done

dmitry-stepanenko commented on libs/platform/src/lib/table/table.service.ts:
```diff
@@ -226,9 +263,126 @@ export class TableService {
+            return;
+        }
+
+        const focusableElements = getFocusableElements(tableCell._elRef.nativeElement);
+
+        if (focusableElements.length) {
+            focusableElements[0].focus();
+            this.setFocusInsideCell();
+        }
```
I think in this case you can have an extra param in `getFocusableElements` to run `querySelector` instead of `querySelectorAll`.

dmitry-stepanenko replied:
Because you need only first entry. `querySelector` is much faster than `querySelectorAll`. 
https://measurethat.net/Benchmarks/Show/5353/0/queryselector-vs-queryselectorall-vs-getelementsbyclass
![image](https://user-images.githubusercontent.com/33101123/145365983-6435f08f-b8b4-4821-8fe3-4c56d5749959.png)

dmitry-stepanenko commented on libs/platform/src/lib/table/table.service.ts:
```diff
@@ -226,9 +263,126 @@ export class TableService {
+            return;
+        }
+
+        const focusableElements = getFocusableElements(tableContainer);
+        const currFocusElementIndex = focusableElements.findIndex((el) => el === document.activeElement);
+
+        if (currFocusElementIndex === 1 && event.shiftKey) {
+            this.removeFocusInsideCell();
+        }
+
+        if (currFocusElementIndex === focusableElements.length - 1 && !event.shiftKey) {
+            focusableMock.focus();
+            this.removeFocusInsideCell();
+        }
```
Iterating over the set of focusable elements in the entire table each time focus is changed is quite inefficient. Moreover you're limiting the focus ability only to certain elements `a, button, input, textarea, select, details, [tabindex]:not([tabindex="-1"]`. I'm wondering if it's a good idea..

What exactly this logic does? I see that this is being invoked specifically in the case when `TAB` is hit.

platon-rov replied:
The idea is to have the next behavior: jump over the cells with arrow keys, use f2 to go in/out of the cell, use tab to move between focusable elements in the table. We tried to reach the same behavior as the sapui5 grid table has.

`a, button, input, textarea, select, details, [tabindex]:not([tabindex="-1"]` are the selectors of all possible focusable elements. The opposite way is to bother users to apply the certain directive to focusable elements in the table cells.

dmitry-stepanenko replied:
I see. Maybe we should then at least add a custom attribute that developers can use to specify the element is focusable?

platon-rov replied:
If user needs focusable `div` or something he can add `tabindex=0` to it, so it will be catched and added to the tabbable elements, because of the `[tabindex]:not([tabindex="-1"]` rule.

platon-rov replied:
About what you sent, they're also checking all that stuff but just in more progressive way and also checking iframe, video & audio elements because of their controls and so on.

So maybe we need to improve our approach, but as a bare minimum rules above is enough i think.

dimamarksman commented on libs/platform/src/lib/table/table.component.html:
```diff
@@ -338,21 +366,27 @@
                                 [class.fd-table__cell--expand]="_isTreeRowFirstCell(colIdx, row)"
                                 [ngStyle]="_getCellStyles(column)"
                                 [attr.aria-expanded]="_isTreeRow(row) ? row.expanded : null"
+                                attr.aria-label="Row {{ rowIdx }} of {{ _totalItems }}, Column {{ colIdx }} of {{
+                                    _tableColumnsLength
+                                }}. "
```
shouldn't we add i18n marker?

platon-rov replied:
I would also suggest to move it to getter.

aberikashvili replied:
Added

dimamarksman commented on libs/platform/src/lib/table/table.component.ts:
```diff
@@ -1979,6 +1992,52 @@ export class TableComponent<T = any> extends Table implements AfterViewInit, OnD
+                    this._tableService.focusNextTableCell(FIRST_CELL_NAVIGATION_ID);
+                }
+
+                shouldDelegateFocus = true;
```
as far as I can see `shouldDelegateFocus` is always true, don't understand why we need it

aberikashvili replied:
removed

dimamarksman commented:
can not move up via arrows keys in this case

![image](https://user-images.githubusercontent.com/3263744/144858019-234c4ac6-4815-4c1e-9d5e-beb10c4ad726.png)

But can go from top to bottom

mikerodonnell89 commented on libs/platform/src/lib/components/table/table.component.scss:
```diff
@@ -0,0 +1,416 @@
+$fd-table-border: 0.0625rem solid #e4e4e4;
```
I believe this file was added accidentally

aberikashvili replied:
@mikerodonnell89 why?

mikerodonnell89 replied:
It is in a directory by itself
<img width="357" alt="Screen Shot 2021-12-16 at 9 40 24 AM" src="https://user-images.githubusercontent.com/2471874/146392214-9f01617d-0c3a-4738-b086-270ff6339c1e.png">

And the first few lines are essentially a duplicate of [this scss file](https://github.com/SAP/fundamental-ngx/blob/0a69acfa9595e168458db82e5f605c88994cc634/libs/platform/src/lib/table/table.component.scss). Note the almost identical pathnames except this file one contains the directory "components".  Should the changes here actually be in [this file](https://github.com/SAP/fundamental-ngx/blob/0a69acfa9595e168458db82e5f605c88994cc634/libs/platform/src/lib/table/table.component.scss) or is it intentional to have these styles in their own directory?
 
Same question for the constants/index.ts file above

aberikashvili commented:
> 

Fixed, please check again @dimamarksman

aberikashvili commented:
@mikerodonnell89 your comment is addressed

dmitry-stepanenko reviewed (commented):
@aberikashvili lgtm, left one small comment.

Also I'm seeing that these 2 comments were not addressed:
- https://github.com/SAP/fundamental-ngx/pull/5692#discussion_r762889143
- https://github.com/SAP/fundamental-ngx/pull/5692#discussion_r762946919

mikerodonnell89 reviewed (changes requested):
regarding issue #3841 - Chrome screenreader announces "row 0 of 16" and I think it should be 1-indexed, i.e. 1 of 16

aberikashvili commented on libs/platform/src/lib/table/table.component.html:
```diff
@@ -318,11 +341,17 @@
+                                        (keydown.space)="_toggleSelectableRow(row, $event)"
                                     >
                                         <fd-checkbox
+                                            attr.aria-label="Select row number {{ rowIdx + 1 }} of {{ _totalItems }}"
```
added

Veerakarthick commented:
Hello colleague,

The reported issues are the duplicate tickets which has been already mentioned in different PR.

Regards,
Karthikeyan.c

aberikashvili commented:
> Two issues:
> 
> 1. Chrome screenreader announces "row 0 of 16" and "column 0 of 4" and I think it should be 1-indexed, i.e. 1 of 16
> 2. When checking a checkbox, the Chrome screenreader announces "Checkbox checked" several times

Hey @mikerodonnell89, I don't see the behavior you mentioned in MacOS voiceover, why are you checking in Chrome Screenreader? did I miss something?

aberikashvili commented:
> Hello colleague,
> 
> The reported issues are the duplicate tickets which has been already mentioned in different PR.
> 
> Regards, Karthikeyan.c

Hey @Veerakarthick , what do you mean?

Veerakarthick commented:
Hello Colleague,

Same issue persist twice in the Accessibility issue dashboard, so i have mentioned one of the ticket as duplicate and will track this issue in one ticket.

Thanks and Regards,
Karthikeyan.c

mikerodonnell89 commented:
Yep can confirm, using the chrome screenreader, when I check a checkbox in the “multiple row selection” example, the screenreader announces “checkbox checked” 2 or 3 times. Would be nice to see this fixed, however I know Chrome is a lower priority on our list of screenreader compatibility so we can move forward

Veerakarthick commented:
Hello Colleague,

The  PR is associated with below tickets and the tickets need Rework so am keeping this PR open with Rework required  lable.

Associated Tickets:
Closes https://github.com/SAP/fundamental-ngx/issues/3841
Closes https://github.com/SAP/fundamental-ngx/issues/4962
Closes https://github.com/SAP/fundamental-ngx/issues/4668


Thanks and Regards,
Karthikeyan.c
